File: src/main/java/ludwig/script/Parser.java
package ludwig.script;

import ludwig.changes.Change;
import ludwig.changes.InsertNode;
import ludwig.changes.InsertReference;
import ludwig.interpreter.ClassType;
import ludwig.model.*;
import ludwig.workspace.Workspace;
import org.pcollections.HashPMap;
import org.pcollections.HashTreePMap;

import java.io.IOException;
import java.io.Reader;
import java.util.Collections;
import java.util.List;

import static ludwig.utils.NodeUtils.isField;

public class Parser {
    private final List<String> tokens;
    private final Workspace workspace;

    private int pos;
    private HashPMap<String, NamedNode> locals = HashTreePMap.empty();
    private Signature superFunction;


    private Parser(List<String> tokens, Workspace workspace) {
        this.tokens = tokens;
        this.workspace = workspace;
    }


    public static void parse(Reader reader, Workspace workspace, ProjectNode projectNode) throws ParserException, IOException, LexerException {
        parse(Lexer.read(reader), workspace, projectNode);
    }

    public static void parse(List<String> tokens, Workspace workspace, ProjectNode projectNode) throws ParserException {
        new Parser(tokens, workspace).parse(projectNode);
    }

    public static NamedNode item(PackageNode findByName, String name) {
        return findByName.children().stream().filter(n -> n instanceof NamedNode).map(n -> (NamedNode) n).filter(it -> it.name().equals(name)).findFirst().orElse(null);
    }

    private void parse(ProjectNode projectNode) throws ParserException {
        PackageNode packageNode = parseSignatures(projectNode);
        parseBodies(packageNode);
    }

    private PackageNode parseSignatures(ProjectNode projectNode) throws ParserException {
        consume("(");
        consume("package");

        String packageName = nextToken();
        PackageNode packageNode = projectNode.children().stream().map(n -> (PackageNode) n)
                .filter(n -> n.name().equals(packageName))
                .findFirst()
                .orElseGet(() -> append(projectNode, new PackageNode().name(packageName)));

        consume(")");


        while (pos < tokens.size()) {
            parseSignature(packageNode);
        }

        return packageNode;
    }

    private void parseSignature(PackageNode packageNode) throws ParserException {
        consume("(");

        switch (nextToken()) {
            case "class": {
                ClassNode classNode = append(packageNode, new ClassNode().name(nextToken()));
                if (!currentToken().equals(")")) {
                    ClassNode superClass = (ClassNode) find(nextToken());
                    append(classNode, new ReferenceNode(superClass));
                }
                while (!currentToken().equals(")")) {
                    append(classNode, new VariableNode().name(nextToken()));
                }
                consume(")");
                ClassType.of(classNode);
                break;
            }
            case "def": {
                boolean lazy = false;
                if (currentToken().equals("lazy")) {
                    lazy = true;
                    consume("lazy");
                }
                FunctionNode fn = append(packageNode, new FunctionNode().name(nextToken()).setLazy(lazy));
                while (!currentToken().equals(")")) {
                    append(fn, new VariableNode().name(nextToken()));
                }
                consume(")");

                skipBody();
                break;
            }
            case "method": {
                while (!nextToken().equals(")")) ;
                skipBody();
                break;
            }
        }
    }

    private void skipBody() throws ParserException {
        consume("(");
        int level = 1;
        while (level != 0 && pos < tokens.size()) {
            switch (nextToken()) {
                case "(":
                    level++;
                    break;
                case ")":
                    level--;
                    break;
            }
        }
    }

    private void parseBodies(PackageNode packageNode) throws ParserException {
        rewind();
        consume("(");
        consume("package");

        nextToken();
        consume(")");

        while (pos < tokens.size()) {
            parseBody(packageNode);
        }
    }

    private void parseBody(PackageNode packageNode) throws ParserException {
        consume("(");
        switch (nextToken()) {
            case "class":
                while (!nextToken().equals(")")) ;
                break;
            case "def": {
                if (currentToken().equals("lazy")) {
                    consume("lazy");
                }
                FunctionNode node = (FunctionNode) item(packageNode, nextToken());
                locals = HashTreePMap.empty();
                for (Node child : node.children()) {
                    if (!(child instanceof VariableNode)) {
                        break;
                    }
                    VariableNode variableNode = (VariableNode) child;
                    locals = locals.plus(variableNode.name(), variableNode);
                }
                while (!nextToken().equals(")")) ;
                consume("(");
                while (pos < tokens.size() && !currentToken().equals(")")) {
                    parseChild(node);
                }

                if (pos < tokens.size()) {
                    nextToken();
                }
                break;
            }
            case "method": {
                ClassNode classNode = (ClassNode) find(nextToken());
                FunctionNode fn = (FunctionNode) find(nextToken());
                OverrideNode node = append(packageNode, new OverrideNode());
                append(node, new ReferenceNode(fn));

                superFunction = findSuper(classNode, fn);


                locals = HashTreePMap.empty();

                for (Node child : fn.children()) {
                    if (!(child instanceof VariableNode)) {
                        break;
                    }
                    VariableNode variableNode = (VariableNode) child;
                    locals = locals.plus(variableNode.name(), variableNode);
                }

                while (!nextToken().equals(")")) ;
                consume("(");
                while (pos < tokens.size() && !currentToken().equals(")")) {
                    parseChild(node);
                }

                if (pos < tokens.size()) {
                    nextToken();
                }

                ClassType.of(classNode).overrides().put(fn, node);
                break;
            }
            case "field":
                nextToken();
                consume(")");
                break;

        }
    }

    private Signature findSuper(ClassNode classNode, FunctionNode fn) {
        ClassType t = ClassType.of(classNode);
        Signature s = t.implementation(fn);

        while (t != null) {
            Signature s1 = t.implementation(fn);
            if (s1 != s) {
                return s1;
            }

            t = t.superClass();
        }
        return fn;
    }


    private void parseChild(Node parent) throws ParserException {
        int level = 0;
        while (currentToken().equals("(")) {
            level++;
            nextToken();
        }

        try {
            String head = nextToken();

            switch (head) {
                case "call":
                case "if":
                case "else":
                case "return":
                case "throw":
                case "list":
                case "break":
                case "continue": {
                    Node node = append(parent, createSpecial(head));
                    while (!currentToken().equals(")")) {
                        parseChild(node);
                    }
                    break;
                }
                case "ref":
                    FunctionReferenceNode ref = append(parent, new FunctionReferenceNode());
                    append(ref, new ReferenceNode(find(nextToken())));
                    break;
                case "for": {
                    ForNode node = append(parent, new ForNode());
                    VariableNode var = new VariableNode();
                    var.name(nextToken());
                    append(node, var);

                    HashPMap savedLocals = locals;
                    locals = locals.plus(var.name(), var);

                    while (!currentToken().equals(")")) {
                        parseChild(node);
                    }
                    locals = savedLocals;
                    break;
                }
                case "=": {
                    String name = nextToken();
                    AssignmentNode node = append(parent, new AssignmentNode());

                    boolean isField = false;
                    int savedPos = pos;

                    NamedNode f = find(name);
                    if (isField(f)) {
                        ReferenceNode r = append(node, new ReferenceNode(f));
                        parseChild(r);
                        if (!currentToken().equals(")")) {
                            parseChild(node);
                            isField = true;
                        }
                    }

                    if (!isField) {
                        pos = savedPos;
                        node.children().clear();
                        if (locals.containsKey(name)) {
                            append(node, new ReferenceNode(locals.get(name)));
                            parseChild(node);
                        } else {
                            VariableNode lhs = append(node, new VariableNode().name(name));
                            locals = locals.plus(name, lhs);
                            parseChild(node);
                        }
                    }
                    break;
                }
                case "λ":
                case "\\": {
                    LambdaNode node = append(parent, new LambdaNode());
                    HashPMap<String, NamedNode> savedLocals = locals;
                    while (!currentToken().equals(")")) {
                        VariableNode param = append(node, new VariableNode().name(nextToken()));
                        locals = locals.plus(param.name(), param);
                    }
                    consume(")");
                    consume("(");
                    while (!currentToken().equals(")")) {
                        parseChild(node);
                    }
                    locals = savedLocals;
                    break;
                }

                default: {
                    Node headNode = "super".equals(head) ? (Node) superFunction : find(head);

                    if (isField(headNode)) {
                        int savedPos = pos;
                        VariableNode fn = (VariableNode) headNode;
                        ReferenceNode r = append(parent, new ReferenceNode(fn));
                        if (currentToken().equals(")")) {
                            pos = savedPos;
                            parent.children().remove(parent.children().size() - 1);
                        } else {
                            parseChild(r);
                            return;
                        }
                    }

                    if (locals.containsKey(head)) {
                        append(parent, new ReferenceNode(locals.get(head)));
                    } else if (headNode instanceof FunctionNode) {
                        FunctionNode fn = (FunctionNode) headNode;
                        ReferenceNode r = append(parent, new ReferenceNode(fn));
                        for (Node param : fn.children()) {
                            if (!(param instanceof VariableNode)) {
                                break;
                            }
                            parseChild(r);
                        }
                    } else if (headNode instanceof OverrideNode) {
                        FunctionNode fn = (FunctionNode) ((ReferenceNode) headNode.children().get(0)).ref();
                        ReferenceNode r = append(parent, new ReferenceNode(headNode));
                        for (Node param : fn.children()) {
                            if (!(param instanceof VariableNode)) {
                                break;
                            }
                            parseChild(r);
                        }
                    } else if (headNode instanceof ClassNode) {
                        ClassNode cn = (ClassNode) headNode;
                        ReferenceNode r = append(parent, new ReferenceNode(cn));
                        while (!currentToken().equals(")")) {
                            parseChild(r);
                        }
                    } else if (Lexer.isLiteral(head)) {
                        append(parent, new LiteralNode(head));
                    } else {
                        throw new ParserException("Unknown symbol: " + head);
                    }
                }
            }
        } finally {
            for (int i = 0; i < level; i++) {
                consume(")");
            }
        }
    }

    private String nextToken() {
        return tokens.get(pos++);
    }

    private String currentToken() {
        return tokens.get(pos);
    }

    private void consume(String token) throws ParserException {
        if (!nextToken().equals(token)) {
            throw new ParserException("Expected " + token);
        }
    }

    private void rewind() {
        pos = 0;
    }

    // TODO: Optimize
    private NamedNode find(String name) {
        for (ProjectNode project : workspace.getProjects()) {
            for (Node node : project.children()) {
                PackageNode packageNode = (PackageNode) node;
                if (item(packageNode, name) != null) {
                    return item(packageNode, name);
                }
            }
        }
        return null;
    }

    private Node createSpecial(String token) {
        switch (token) {
            case "call":
                return new CallNode();
            case "if":
                return new IfNode();
            case "else":
                return new ElseNode();
            case "return":
                return new ReturnNode();
            case "list":
                return new ListNode();
            case "throw":
                return new ThrowNode();
            case "break":
                return new BreakNode();
            case "continue":
                return new ContinueNode();
        }
        return null;
    }

    private <T extends Node> T append(Node<?> parent, T node) {
        if (node instanceof ReferenceNode) {
            InsertReference change = new InsertReference()
                    .id(Change.newId())
                    .parent(parent.id())
                    .prev(parent.children().isEmpty() ? null : parent.children().get(parent.children().size() - 1).id())
                    .next(null)
                    .ref(((ReferenceNode) node).ref().id());
            workspace.apply(Collections.singletonList(change));
            return workspace.node(change.id());
        } else {
            if (node instanceof NamedNode) {
                node.id(parent.id() + ":" + ((NamedNode) node).name());
            } else {
                node.id(Change.newId());
            }
            InsertNode change = new InsertNode()
                    .node(node)
                    .parent(parent.id())
                    .prev(parent.children().isEmpty() ? null : parent.children().get(parent.children().size() - 1).id());
            workspace.apply(Collections.singletonList(change));
            return workspace.node(node.id());
        }

    }
}

File: src/main/java/ludwig/script/Parser.kt
package ludwig.script

import ludwig.changes.Change
import ludwig.changes.InsertNode
import ludwig.changes.InsertReference
import ludwig.interpreter.ClassType
import ludwig.model.*
import ludwig.utils.NodeUtils.isField
import ludwig.workspace.Workspace
import org.pcollections.HashPMap
import org.pcollections.HashTreePMap
import java.io.IOException
import java.io.Reader

class Parser private constructor(private val tokens: List<String>, private val workspace: Workspace) {

    private var pos: Int = 0
    private var locals: HashPMap<String, NamedNode<*>> = HashTreePMap.empty()
    private var superFunction: Node<*>? = null

    @Throws(ParserException::class)
    private fun parse(projectNode: ProjectNode) {
        val packageNode = parseSignatures(projectNode)
        parseBodies(packageNode)
    }

    @Throws(ParserException::class)
    private fun parseSignatures(projectNode: ProjectNode): PackageNode {
        consume("(")
        consume("package")

        val packageName = nextToken()
        val packageNode = projectNode.children().stream().map { n -> n as PackageNode }
                .filter { n -> n.name() == packageName }
                .findFirst()
                .orElseGet { append(projectNode, PackageNode().name(packageName)) }

        consume(")")


        while (pos < tokens.size) {
            parseSignature(packageNode)
        }

        return packageNode
    }

    @Throws(ParserException::class)
    private fun parseSignature(packageNode: PackageNode) {
        consume("(")

        when (nextToken()) {
            "class" -> {
                val classNode = append(packageNode, ClassNode().name(nextToken()))
                if (currentToken() != ")") {
                    val superClass = find(nextToken()) as ClassNode?
                    appendRef(classNode, superClass)
                }
                while (currentToken() != ")") {
                    append(classNode, VariableNode().name(nextToken()))
                }
                consume(")")
                ClassType.of(classNode!!)
            }
            "def" -> {
                var lazy = false
                if (currentToken() == "lazy") {
                    lazy = true
                    consume("lazy")
                }
                val fn = append(packageNode, FunctionNode().name(nextToken()).lazy(lazy))
                while (currentToken() != ")") {
                    append(fn, VariableNode().name(nextToken()))
                }
                consume(")")

                skipBody()
            }
            "method" -> {
                while (nextToken() != ")");
                skipBody()
            }
        }
    }

    @Throws(ParserException::class)
    private fun skipBody() {
        consume("(")
        var level = 1
        while (level != 0 && pos < tokens.size) {
            when (nextToken()) {
                "(" -> level++
                ")" -> level--
            }
        }
    }

    @Throws(ParserException::class)
    private fun parseBodies(packageNode: PackageNode) {
        rewind()
        consume("(")
        consume("package")

        nextToken()
        consume(")")

        while (pos < tokens.size) {
            parseBody(packageNode)
        }
    }

    @Throws(ParserException::class)
    private fun parseBody(packageNode: PackageNode) {
        consume("(")
        when (nextToken()) {
            "class" -> while (nextToken() != ")");
            "def" -> {
                if (currentToken() == "lazy") {
                    consume("lazy")
                }
                val node = item(packageNode, nextToken()) as FunctionNode?
                locals = HashTreePMap.empty<String, NamedNode<*>>()
                for (child in node!!.children()) {
                    if (child !is VariableNode) {
                        break
                    }
                    locals = locals.plus(child.name()!!, child)
                }
                while (nextToken() != ")");
                consume("(")
                while (pos < tokens.size && currentToken() != ")") {
                    parseChild(node)
                }

                if (pos < tokens.size) {
                    nextToken()
                }
            }
            "method" -> {
                val classNode = find(nextToken()) as ClassNode
                val fn = find(nextToken()) as FunctionNode
                val node = append(packageNode, OverrideNode())
                appendRef(node, fn)

                superFunction = findSuper(classNode, fn)


                locals = HashTreePMap.empty<String, NamedNode<*>>()

                for (child in fn!!.children()) {
                    if (child !is VariableNode) {
                        break
                    }
                    locals = locals.plus(child.name()!!, child)
                }

                for (child in classNode!!.children()) {
                    if (child !is VariableNode) {
                        continue
                    }
                    locals = locals.plus(child.name()!!, child)
                }

                while (nextToken() != ")");
                consume("(")
                while (pos < tokens.size && currentToken() != ")") {
                    parseChild(node)
                }

                if (pos < tokens.size) {
                    nextToken()
                }

                ClassType.of(classNode).overrides().put(fn, node!!)
            }
            "field" -> {
                nextToken()
                consume(")")
            }
        }
    }

    private fun findSuper(classNode: ClassNode, fn: FunctionNode): Node<*> {
        var t: ClassType? = ClassType.of(classNode)
        val s = t!!.implementation(fn)

        while (t != null) {
            val s1 = t.implementation(fn)
            if (s1 !== s) {
                return s1
            }

            t = t.superClass()
        }
        return fn
    }


    @Throws(ParserException::class)
    private fun parseChild(parent: Node<*>?) {
        var level = 0
        while (currentToken() == "(") {
            level++
            nextToken()
        }

        try {
            val head = nextToken()

            when (head) {
                "call", "if", "else", "return", "throw", "try", "catch", "list", "break", "continue" -> {
                    val node = append(parent, createSpecial(head))
                    while (currentToken() != ")") {
                        parseChild(node)
                    }
                }
                "ref" -> {
                    val ref = append(parent, FunctionReferenceNode())
                    appendRef(ref, find(nextToken()))
                }
                "for" -> {
                    val node = append(parent, ForNode())
                    val `var` = VariableNode()
                    `var`.name(nextToken())
                    append(node, `var`)

                    val savedLocals = locals
                    locals = locals.plus(`var`.name()!!, `var`)

                    while (currentToken() != ")") {
                        parseChild(node)
                    }
                    locals = savedLocals
                }
                "=" -> {
                    val name = nextToken()
                    val node = append(parent, AssignmentNode())

                    var isField = false
                    val savedPos = pos

                    val f = find(name)
                    if (isField(f)) {
                        val r = appendRef(node, f)
                        parseChild(r)
                        if (currentToken() != ")") {
                            parseChild(node)
                            isField = true
                        }
                    }

                    if (!isField) {
                        pos = savedPos
                        node!!.children().clear()
                        if (locals.containsKey(name)) {
                            appendRef(node, locals[name])
                            parseChild(node)
                        } else {
                            val lhs = append(node, VariableNode().name(name))
                            locals = locals.plus(name, lhs)
                            parseChild(node)
                        }
                    }
                }
                "λ", "\\" -> {
                    val node = append(parent, LambdaNode())
                    val savedLocals = locals
                    while (currentToken() != ")") {
                        val param = append(node, VariableNode().name(nextToken()))
                        locals = locals.plus(param!!.name()!!, param)
                    }
                    consume(")")
                    consume("(")
                    while (currentToken() != ")") {
                        parseChild(node)
                    }
                    locals = savedLocals
                }

                else -> {
                    if (locals.containsKey(head)) {
                        val local = locals[head]
                        if (isField(local)) {
                            val savedPos = pos
                            val fn = local as VariableNode
                            val r = appendRef(parent, fn)
                            if (currentToken() == ")") {
                                pos = savedPos
                                parent!!.children().removeAt(parent.children().size - 1)
                            } else {
                                parseChild(r)
                                return
                            }
                        } else {
                            appendRef(parent, local)
                        }
                        return
                    }

                    val headNode = if ("super" == head) superFunction else find(head)
                    if (headNode is FunctionNode) {
                        val fn = headNode as FunctionNode?
                        val r = appendRef(parent, fn)
                        for (param in fn!!.children()) {
                            if (param !is VariableNode) {
                                break
                            }
                            parseChild(r)
                        }
                    } else if (headNode is OverrideNode) {
                        val fn = (headNode.children()[0] as ReferenceNode).ref() as FunctionNode
                        val r = appendRef(parent, headNode)
                        for (param in fn.children()) {
                            if (param !is VariableNode) {
                                break
                            }
                            parseChild(r)
                        }
                    } else if (headNode is ClassNode) {
                        val cn = headNode as ClassNode?
                        val r = appendRef(parent, cn)
                        while (currentToken() != ")") {
                            parseChild(r)
                        }
                    } else if (Lexer.isLiteral(head)) {
                        append(parent, LiteralNode(head))
                    } else {
                        throw ParserException("Unknown symbol: " + head)
                    }
                }
            }
        } finally {
            for (i in 0 until level) {
                consume(")")
            }
        }
    }

    private fun nextToken(): String {
        return tokens[pos++]
    }

    private fun currentToken(): String {
        return tokens[pos]
    }

    @Throws(ParserException::class)
    private fun consume(token: String) {
        if (nextToken() != token) {
            throw ParserException("Expected " + token)
        }
    }

    private fun rewind() {
        pos = 0
    }

    // TODO: Optimize
    private fun find(name: String): NamedNode<*>? {
        return workspace.projects
                .flatMap { it.children() }
                .map { it as PackageNode }
                .firstOrNull { item(it, name) != null }
                ?.let { item(it, name) }
    }

    private fun createSpecial(token: String): Node<*>? {
        when (token) {
            "call" -> return CallNode()
            "if" -> return IfNode()
            "else" -> return ElseNode()
            "return" -> return ReturnNode()
            "list" -> return ListNode()
            "throw" -> return ThrowNode()
            "try" -> return TryNode()
            "catch" -> return ClassNode()
            "break" -> return BreakNode()
            "continue" -> return ContinueNode()
        }
        return null
    }

    private fun <T : Node<*>> append(parent: Node<*>?, node: T?): T? {
        if (node is NamedNode<*>) {
            node.id(parent!!.id() + ":" + (node as NamedNode<*>).name())
        } else {
            node!!.id(Change.newId())
        }
        val change = InsertNode()
                .node(node)
                .parent(parent!!.id())
                .prev(if (parent.children().isEmpty()) null else parent.children()[parent.children().size - 1].id())
        workspace.apply(listOf(change))
        return workspace.node<Node<*>>(node.id()) as T?
    }

    private fun appendRef(parent: Node<*>?, node: Node<*>?): ReferenceNode {
        val change = InsertReference()
                .id(Change.newId())
                .parent(parent!!.id())
                .prev(if (parent.children().isEmpty()) null else parent.children()[parent.children().size - 1].id())
                .next(null)
                .ref(node!!.id()!!)
        workspace.apply(listOf(change))
        return workspace.node<Node<*>>(change.id()) as ReferenceNode
    }

    companion object {


        @Throws(ParserException::class, IOException::class, LexerException::class)
        fun parse(reader: Reader, workspace: Workspace, projectNode: ProjectNode) {
            parse(Lexer.read(reader), workspace, projectNode)
        }

        @Throws(ParserException::class)
        fun parse(tokens: List<String>, workspace: Workspace, projectNode: ProjectNode) {
            Parser(tokens, workspace).parse(projectNode)
        }

        fun item(findByName: PackageNode, name: String): NamedNode<*>? {
            return findByName.children().stream().filter { n -> n is NamedNode<*> }.map { n -> n as NamedNode<*> }.filter { it -> it.name() == name }.findFirst().orElse(null)
        }
    }
}
